src/lib.rs

```diff
@@ -869,34 +866,40 @@ mod unix {
                 return Ok(());
             }
 
-            state.on_checkpoint(self.ptr, self.len)
+            state.on_checkpoint(self.address, self.len, self.page_size)
         }
     }
 
     impl PageFaultCheckerState {
-        fn on_checkpoint(&mut self, ptr: *mut usize, len: usize) -> Result<(), PageFaultError> {
+        fn on_checkpoint(
+            &mut self,
+            address: usize,
+            len: usize,
+            page_size: usize,
+        ) -> Result<(), RuntimeError> {
             use libc::mincore;
 
-            unsafe {
-                let page_size = get_page_size().unwrap();
-                let base_address = ptr as usize;
-                let mem_size = len * size_of::<usize>();
-
-                let aligned_address = base_address & !(page_size - 1);
-                let aligned_ptr = (aligned_address as *mut usize).cast();
-                let aligned_mem_size = mem_size + (base_address - aligned_address);
-
-                // buf needs to be at least (aligned_mem_size + page_size - 1) / page_size
-                let mut buf: Vec<u8> = vec![0; aligned_mem_size.div_ceil(page_size)];
-                assert_eq!(
-                    mincore(aligned_ptr, aligned_mem_size, buf.as_mut_ptr().cast()),
-                    0,
-                    "mincore did not return 0"
-                );
-                if buf.iter().any(|n| *n == 0) {
-                    trace!("Detected page fault");
-                    return Err(PageFaultError);
-                }
+            let mem_size = len * size_of::<usize>();
+
+            let aligned_address = address / page_size * page_size;
+            let aligned_mem_size = mem_size + (address - aligned_address);
+
+            // buf needs to be at least (aligned_mem_size + page_size - 1) / page_size
+            let mut buf: Vec<u8> = vec![0; aligned_mem_size.div_ceil(page_size)];
+            assert_eq!(
+                unsafe {
+                    mincore(
+                        (aligned_address as *mut usize).cast(),
```

Comment:
nit: Just use [`std::ffi::c_void`](https://doc.rust-lang.org/std/ffi/enum.c_void.html) to avoid having to cast.